rfcs/015-migrate-api-to-esm.md

```diff
@@ -0,0 +1,157 @@
+  ```
+
+- `module` is undefined in ESM. Scripts that can be both imported and executed, need to check this differently.
+
+  ```js
+  import { pathToFileURL } from 'url';
+
+  export function run() {
+    // some function
+  }
+
+  // is being executed as a script
+  if (import.meta.url === pathToFileURL(process.argv[1]).href) {
+    run();
+  }
+  ```
+
+- `require` is undefined. Scripts that need `required` need to be refactored.
+
+  ```js
+  import { createRequire } from 'node:module';
+  const require = createRequire(import.meta.url);
+  // eslint-disable-next-line import/no-commonjs
+  const cloudflareIps = require('cloudflare-ip/ips.json');
+  ```
+
+- ESM imports are immutable, test libraries for mocking and stubbing imports do not work with ESM. `sinon` would have to be replaced by an alternative and tests refactored.
+
+  - [testdouble](https://github.com/testdouble/testdouble.js) is an alternative that plugs into NodeJs module loader to provide this function.
+
+    ```js
+    beforeEach(async () => {
+      const transferWiseLibMock = await td.replaceEsm('../../../server/lib/transferwise.js');
+      // getTransfer is a testdouble that can be used to stub and verify calls to it.
+      getTransfer = transferWiseLibMock.getTransfer;
+
+      // cron is the test subject, imported after replacing the ESM dependency above.
+      const cron = await import('../../../cron/daily/check-pending-transferwise-transactions.js');
+      checkPendingTransfers = cron.run;
+    });
+
+    it('test', () => {
+      td.when(
+          getTransfer(
+            td.matchers.anything(), 
+            td.matchers.anything(),
+          )
+      ).thenResolve({
+        status: 'outgoing_payment_sent',
+        id: 1234,
+      });
+    });
+    ```
+
+- `babel-loader` needs to be replaced by [ts-node](https://github.com/TypeStrong/ts-node) in order to execute the development live-reload API and execute crons. `ts-node` supports transpiling TypeScript with ESM support.
+
+  ! This means dropping babel transpilation. This is not an issue at this moment as all features used in source are currently supported by our NodeJS version.
+
+  ```json
+  "script": "ts-node-esm $1",
+  ```
+
+- `sequelize` imports migration files using `require`, meaning they are imported as `CommonJS`. To deal with this, migrations would have to be defined with a `.cts` or `.cjs` extension to indicate CommonJS format to the loader. This also means that importing server code from migrations (something that we do sometimes) would require async imports to cross the CommonJS and EMS boundary.
+
+## Unknowns
+
+With the exception of tests (unit and end-to-end), these migrations have been tested, code compiles and executes. However, since tests were not migrated, verification of runtime changes is still pending.
+
+## Alternatives
+
+Don't migrate to ESM. It would still be possible to use ESM modules and keep up to date with the ecosystem by making use of async imports.
```

Comment:
Do you think it will be easier to migrate later on when more libraries/tooling will move to ESM, or is it mature already and there would be no difference if we delay the decision?

Reply:
I think its safe to progressively migrate now. The tooling required is moving from `babel-loader`/`babel` to `ts-node`/`tsc`, and we can make it without changing the project module type. After we have progressed on these changes which are compatible with both ESM/CommonJS we can see about flipping the switch finally.